File: backend/app/schemas/verification.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

Verdict = Literal["SUPPORTED", "CONTRADICTED", "NOT_FOUND", "NEEDS_REVIEW"]


class ModelVerification(BaseModel):
    verdict: Verdict
    confidence: float = Field(ge=0, le=1)
    quote: str
    explanation: str = Field(min_length=3, max_length=1000)
    page_number: int | None = None
    chunk_id: str | None = None


class SourceResponse(BaseModel):
    document_id: str
    document_name: str
    page_number: int | None
    chunk_id: str | None
    start_char: int | None
    end_char: int | None


class ChecksResponse(BaseModel):
    quote_verified: bool
    numbers_consistent: bool
    percentages_consistent: bool
    dates_consistent: bool
    currency_consistent: bool


class MetricsResponse(BaseModel):
    latency_ms: int
    chunks_searched: int
    chunks_retrieved: int
    context_characters: int
    provider: str
    model: str
    prompt_version: str


class VerificationResponse(BaseModel):
    verification_id: str
    claim_id: str
    verdict: Verdict
    confidence: float
    quote: str
    explanation: str
    source: SourceResponse
    checks: ChecksResponse
    metrics: MetricsResponse
    created_at: datetime


class BulkVerifyRequest(BaseModel):
    claim_ids: list[str] | None = None

File: frontend/src/api/verification.ts
import { z } from 'zod'
import { apiMutation } from './client'

export const verificationSchema=z.object({verification_id:z.string(),claim_id:z.string(),verdict:z.enum(['SUPPORTED','CONTRADICTED','NOT_FOUND','NEEDS_REVIEW']),confidence:z.number(),quote:z.string(),explanation:z.string(),source:z.object({document_id:z.string(),document_name:z.string(),page_number:z.number().nullable(),chunk_id:z.string().nullable(),start_char:z.number().nullable(),end_char:z.number().nullable()}),checks:z.object({quote_verified:z.boolean(),numbers_consistent:z.boolean(),percentages_consistent:z.boolean(),dates_consistent:z.boolean(),currency_consistent:z.boolean()}),metrics:z.object({latency_ms:z.number(),chunks_searched:z.number(),chunks_retrieved:z.number(),context_characters:z.number(),provider:z.string(),model:z.string(),prompt_version:z.string()}),created_at:z.string()})
export const verifyClaim=(id:string)=>apiMutation(`/claims/${id}/verify`,verificationSchema)
export const verifyClaims=(documentId:string,claimIds?:string[])=>apiMutation(`/documents/${documentId}/verify`,z.array(verificationSchema),'POST',{claim_ids:claimIds})

